homework/ba-nguyen/lesson-8/javascript-4.md

```diff
@@ -0,0 +1,273 @@
+  .then((array) => {
+    console.log("We should not get here");
+  })
+  .catch((error) => {
+    if (error != "X") {
+      console.log("Unexpected failure:", error);
+    }
+  });
+```
+
+**Question 4**: How can we log the values that are commented out after the console.log statement?
+
+```js
+function* startGame() {
+  const answer = yield "Do you love JavaScript?";
+  if (answer !== "Yes") {
+    return "Oh wow... Guess we're gone here";
+  }
+  return "JavaScript loves you back ❤️";
+}
+
+const game = startGame();
+console.log(/* 1 */); // Do you love JavaScript?
+console.log(/* 2 */); // JavaScript loves you back ❤️
+```
+
+A: game.next("Yes").value and game.next().value <br/>
+B: game.next.value("Yes") and game.next.value() <br/>
+`C: game.next().value and game.next("Yes").value` => Answer <br/>
+D: game.next.value() and game.next.value("Yes")
+
+**Question 5**: What's the output?
+
+```js
+function getItems(fruitList, ...args, favoriteFruit) {
+  return [...fruitList, ...args, favoriteFruit]
+}
+
+getItems(["banana", "apple"], "pear", "orange")
+[banana, apple, p, e, a, r, orange]
+```
+
+A: ["banana", "apple", "pear", "orange"] <br/>
+B: [["banana", "apple"], "pear", "orange"] <br/>
+C: ["banana", "apple", ["pear"], "orange"] <br/>
+`D: SyntaxError` => Answer
+
+Explain: `...` in param of function is only used for the rest of params and `...` is also used for array
+
+**Question 6**: What's the output?
+
+```js
+const getList = ([x, ...y]) => [x, y]
+const getUser = user => { name: user.name, age: user.age }
+
+const list = [1, 2, 3, 4]
+const user = { name: "Lydia", age: 21 }
+
+console.log(getList(list))
+console.log(getUser(user))
+```
+
+`A: [1, [2, 3, 4]] and SyntaxError` => Answer <br/>
+B: [1, [2, 3, 4]] and { name: "Lydia", age: 21 } <br/>
+C: [1, 2, 3, 4] and { name: "Lydia", age: 21 } <br/>
+D: Error and { name: "Lydia", age: 21 }
+
+Explain: function getUser is syntaxError, to fix that:
+
+```js
+const getUser = (user) => ({ name: user.name, age: user.age });
+```
+
+**Question 7**: What's the output?
+
+```js
+console.log(`${((x) => x)("I love")} to program`);
+```
+
+`A: I love to program` => Answer <br/>
+B: undefined to program <br/>
+C: ${(x => x)('I love') to program <br/>
+D: TypeError
+
+**Question 8**: What's the output?
+
+```js
+const person = {
+  name: "Lydia",
+  age: 21,
+};
+
+const changeAge = (x = { ...person }) => (x.age += 1);
+const changeAgeAndName = (x = { ...person }) => {
+  x.age += 1;
+  x.name = "Sarah";
+};
+
+changeAge(person);
+changeAgeAndName();
+
+console.log(person);
+```
+
+A: {name: "Sarah", age: 22} <br/>
+B: {name: "Sarah", age: 23} <br/>
+`C: {name: "Lydia", age: 22}` => Answer <br/>
+D: {name: "Lydia", age: 23}
+
+**Question 9**: What's the output?
+
+```js
+async function* range(start, end) {
+  for (let i = start; i <= end; i++) {
+    yield Promise.resolve(i);
+  }
+}
+
+(async () => {
+  const gen = range(1, 3);
+  for await (const item of gen) {
```

Comment:
@quangh33 could you tag thaycacac in this comment? can he explain for me more about this syntax?

Reply:
you can ping him directly on FB. But I think he is on vacation.

Reply:
@nguyenvanbapc66 The generator function range returns an async object with promises for each item in the range we pass: Promise{1}, Promise{2}, Promise{3}. We set the variable gen equal to the async object, after which we loop over it using a for await ... of loop